server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,294 @@
+     *                      completed immediately then this completion happens in the {@link ThreadContext} of the subscribing thread.
+     */
+    @SuppressWarnings({ "rawtypes" })
+    public final void addListener(ActionListener<T> listener, Executor executor, @Nullable ThreadContext threadContext) {
+        if (tryComplete(ref.get(), listener)) {
+            return;
+        }
+
+        final var wrappedListener = fork(executor, preserveContext(threadContext, listener));
+        var currentValue = ref.compareAndExchange(EMPTY, wrappedListener);
+        if (currentValue == EMPTY) {
+            return;
+        }
+        Cell newCell = null;
+        while (true) {
+            if (tryComplete(currentValue, listener)) {
+                return;
+            }
+            if (currentValue instanceof ActionListener firstListener) {
+                final var tail = new Cell(firstListener, null);
+                currentValue = ref.compareAndExchange(firstListener, tail);
+                if (currentValue == firstListener) {
+                    currentValue = tail;
+                }
+                continue;
+            }
+            if (currentValue instanceof Cell head) {
+                if (newCell == null) {
+                    newCell = new Cell(wrappedListener, head);
+                } else {
+                    newCell.next = head;
+                }
+                currentValue = ref.compareAndExchange(head, newCell);
+                if (currentValue == head) {
+                    return;
+                }
+            } else {
+                assert false : "unexpected witness: " + currentValue;
+            }
+        }
+    }
+
+    @Override
+    public final void onResponse(T result) {
+        setResult(new SuccessResult<T>(result));
+    }
+
+    @Override
+    public final void onFailure(Exception exception) {
+        setResult(new FailureResult(exception, wrapException(exception)));
+    }
+
+    protected Exception wrapException(Exception exception) {
+        return exception;
+    }
+
+    /**
+     * @return {@code true} if and only if this listener has been completed (either successfully or exceptionally).
+     */
+    public final boolean isDone() {
+        return isDone(ref.get());
+    }
+
+    /**
+     * @return the result with which this listener completed successfully, or throw the exception with which it failed.
+     *
+     * @throws AssertionError if this listener is not complete yet and assertions are enabled.
+     * @throws IllegalStateException if this listener is not complete yet and assertions are disabled.
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public final T rawResult() throws Exception {
+        final var refValue = ref.get();
+        if (refValue instanceof SuccessResult result) {
+            return (T) result.result();
+        } else if (refValue instanceof FailureResult result) {
+            throw result.exception();
```

Comment:
Hmm IMO we don't care much about the stack trace of the caller here, it's all about the original exception. And I worry more about making decisions on the exception type and the risk of forgetting to unwrap things where necessary. Note that the exception is wrapped up in the callers anyway.

Reply:
So if its already wrapped up we can throw `ExecutionException` here and the callers can get the cause? Then we're making it explicit (also if its a runtime exception, it will turn it checked, which is what we want). This also matches what other future get methods do thoughout the JCL/codebase

Reply:
I think extra wrapping here would be kinda different from other places in the ES codebase. We are pretty ok with completing listeners using the raw exception from some other call stack. Moreover constructing a wrapper exception (capturing the current call stack) is rather expensive and seems a little pointless if we expect callers just to unwrap it again.

I'm not sure there's a good equivalent in the JCL from which to draw inspiration here. Methods like `Future#get` must wrap exceptions in an `ExecutionException` to distinguish "we were completed with an `InterruptedException`" from "we were interrupted while waiting for completion`. Since there's no other possible exceptions here (and the method is called `rawResult`) I would much prefer to keep its current semantics.

Reply:
In that case could this be made `protected`? It's currently only used from subclasses, and so indicates it's not a general-purpose method.